### ISSUE TITLE
Producer::produce succeeds but producer outq_len gets stuck on 1

### ISSUE THREAD

**Issue by LapinDeLaMort** [edenhill/librdkafka#1982]
Read the FAQ first: https://github.com/edenhill/librdkafka/wiki/FAQ

Description
===========
Producer::produce succeeds but producer outq_len gets stuck on 1

How to reproduce
================
I have copied the example code for the CPP interface.
In the example below, the call to 	producer->outq_len() always returns 1

	RdKafka::Conf *conf = RdKafka::Conf::create(RdKafka::Conf::CONF_GLOBAL);
	RdKafka::Conf *tconf = RdKafka::Conf::create(RdKafka::Conf::CONF_TOPIC);
	int32_t partition = RdKafka::Topic::PARTITION_UA;

	ExampleDeliveryReportCb ex_dr_cb;

	/* Set delivery report callback */
	conf->set("dr_cb", &ex_dr_cb, errstr);

	RdKafka::Producer *producer = RdKafka::Producer::create(conf, errstr);
	RdKafka::Topic *topic = RdKafka::Topic::create(producer, m_topicstr + ".TEST",
		tconf, errstr);

	RdKafka::ErrorCode resp =
		producer->produce(topic, partition,
			RdKafka::Producer::RK_MSG_COPY /* Copy payload */,
			const_cast<char *>(msg.c_str()), msg.size(),
			NULL, NULL);
		std::cerr << "% Produced message (" << msg.size() << " bytes)" <<
		std::endl;

	producer->poll(0);

	while (producer->outq_len() > 0) {
		std::cerr << "Waiting for " << producer->outq_len() << std::endl;
		producer->poll(1000);

**IMPORTANT**: Always try to reproduce the issue on the latest released version (see https://github.com/edenhill/librdkafka/releases), if it can't be reproduced on the latest version the issue has been fixed.

Checklist
=========

Please provide the following information:

 - [x] librdkafka version 'librdkafka-0.11'
 - [ ] Apache Kafka version: `kafka_2.11-1.0.0`
 - [ ] librdkafka client configuration: `Default from example above`
 - [ ] Operating system: `Client Win7 Enterprise`
 - [ ] Operating system: `Server Windows Server 2016 Standard`
 - [ ] Provide logs (with `debug=..` as necessary) from librdkafka
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

Apologies, but the system is not letting me upload the server logs. However, there is no message on the server when we try to publish

**Comment by edenhill** [comment]
Is your delivery report callback triggered?

**Comment by TheLoneStranger** [comment]
Hi, 
No, the callback is not triggered.

class ExampleDeliveryReportCb : public RdKafka::DeliveryReportCb {
public:
	void dr_cb(RdKafka::Message &message) {
	}
};

**Comment by edenhill** [comment]
Maybe stick a printf in there to see when it is triggered.

However, you don't seem to configure any brokers.
You need to provide at least one broker in the `bootstrap.servers` config

**Comment by yandaren** [comment]
I meet the problem too, when I produce a message, the outq_len() just increase. and nerver decrease. The topic create success,  but the message produced seems nerver sent to the server, alway in the sent queue

```
namespace producer_ts {

class my_delivery_report_cb : public RdKafka::DeliveryReportCb {
public:
    void dr_cb(RdKafka::Message& message) override {
        printf("message delivery %d bytes, error:%s, key: %s\n",
            (int32_t)message.len(), message.errstr().c_str(), message.key() ? message.key()->c_str() : "");
    }
};

void producer_test() {
    printf("producer test\n");

    int32_t partition = RdKafka::Topic::PARTITION_UA;

    printf("input brokers list(127.0.0.1:9092;127.0.0.1:9093;127.0.0.1:9094):\n");
    std::string broker_list;

    //std::cin >> broker_list;
    broker_list = "127.0.0.1:9092";

    printf("input partition:");

    //std::cin >> partition;
    partition = 0;

    // config 
    RdKafka::Conf* global_conf = RdKafka::Conf::create(RdKafka::Conf::CONF_GLOBAL);
    RdKafka::Conf* topic_conf = RdKafka::Conf::create(RdKafka::Conf::CONF_TOPIC);

    my_hash_partitioner_cb          hash_partitioner;
    my_event_cb                     event_cb;
    my_delivery_report_cb           delivery_cb;
  

    std::string err_string;
    if (topic_conf->set("partitioner_cb", &hash_partitioner, err_string) != RdKafka::Conf::CONF_OK) {
        printf("set partitioner_cb error: %s\n", err_string.c_str());
        return;
    }

    global_conf->set("metadata.broker.list", broker_list, err_string);
    global_conf->set("event_cb", &event_cb, err_string);
    global_conf->set("dr_cb", &delivery_cb, err_string);
    global_conf->set("debug", "all", err_string);

    //dump_config(global_conf);
    //dump_config(topic_conf);


    // create producer
    RdKafka::Producer* producer = RdKafka::Producer::create(global_conf, err_string);
    if (!producer) {
        printf("failed to create producer, %s\n", err_string.c_str());
        return;
    }

    printf("created producer %s\n", producer->name().c_str());

    std::string topic_name;
    while (true) {

        printf("input topic to create:\n");
        std::cin >> topic_name;

        // create topic
        RdKafka::Topic* topic =
            RdKafka::Topic::create(producer, topic_name, topic_conf, err_string);

        if (!topic) {
            printf("try create topic[%s] failed, %s\n",
                topic_name.c_str(), err_string.c_str());
            return;
        }

        printf(">");
        for (std::string line; run && std::getline(std::cin, line); ) {
            if (line.empty()) {
                producer->poll(0);
                continue;
            }

            if (line == "quit") {
                break;
            }

            RdKafka::ErrorCode res = producer->produce(topic, partition,
                RdKafka::Producer::RK_MSG_COPY,
                (char*)line.c_str(), line.size(), NULL, NULL);

            if (res != RdKafka::ERR_NO_ERROR) {
                printf("produce failed, %s\n", RdKafka::err2str(res).c_str());
            }
            else {
                printf("produced msg, bytes %d\n", (int32_t)line.size());
            }

            // do socket io
            producer->poll(0);

            printf("outq_len: %d\n", producer->outq_len());

            //producer->flush(1000);

            //while (run && producer->outq_len()) {
            //    printf("wait for write queue( size %d) write finish\n", producer->outq_len());
            //    producer->poll(1000);
            //}

            printf(">");
        }

        delete topic;

        if (!run) {
            break;
        }
    }

    run = true;

    while (run && producer->outq_len()) {
        printf("wait for write queue( size %d) write finish\n", producer->outq_len());
        producer->poll(1000);
    }

    delete producer;
}
}
```

the log link this
```
LOG-7-TOPPAR: [thrd:HSH-D-3146.wowshga.internal:9092/0]: xxx:9092/0: kafka_cpp_test [0] 3 message(s) in xmit queue (0 added from partition queue)
LOG-7-TOPPAR: [thrd:HSH-D-3146.wowshga.internal:9092/0]: xxx:9092/0: kafka_cpp_test [0] 3 message(s) in xmit queue (0 added from partition queue)
LOG-7-TOPPAR: [thrd:HSH-D-3146.wowshga.internal:9092/0]: xxx:9092/0: kafka_cpp_test [0] 3 message(s) in xmit queue (0 added from partition queue)
```

***
version:
librdkafka: 0.11.5
kafka: 0.11.03

**Comment by LapinDeLaMort** [comment]
It may be of some help, but I have found our particular problem.
Over the weekend, we moved from a windows Kafka server to a Unix version, and this has solved our problem. It turns out that the windows log was not getting cleared down as it should have been. The process that should have deleted the old logs could not delete the files and therefore the log queue was truly blocked.
The fact that it is always reporting a outq_len of 1 is very consistent, no matter how large the outgoing message is.
I'll do some more tests, but i'm happy to close this as the c/cpp code is behaving as it should
